rfcs/0020-telemetry.md

```diff
@@ -0,0 +1,192 @@
+The UF can be used with any application, regardless of language, which makes it the most flexible way to get metrics into Splunk.
+
+It runs in a [sidecar](https://www.oreilly.com/library/view/designing-distributed-systems/9781491983638/ch02.html) to any container that emits metrics.
+
+To use it, create a metrics index in Splunk and add the sidecar to the `container_definition_json` of your `aws-ecs-container-definition` module:
+```terraform
+container_definition_json = jsonencode([
+    module.your-container.json_map_object,
+    module.sidecar-container.json_map_object
+])
+```
+
+You use the StatsD telemetry reporter highlighted above to emit your metrics in the StatsD format, and the UF will automatically forward them.
+
+#### When you should choose this method
+
+- You maintain applications that aren't written in Elixir or want all of your applications to emit metrics in a standard format.
+
+Every language has a library that emits metrics in the StatsD format; they are straightforward to set up.
+You get a lot of flexibility with very little work by running the UF as a sidecar and using an intermediary standard like StatsD.
+
+#### When you shouldn't choose this method
+
+- You only need metrics from Elixir applications.
+- You are hesitant to change your infrastructure.
+- You are concerned about your ECS resource usage.
+
+The UF sidecar uses CPU and memory resources.
+Though the footprint is small, you might be worried that increasing your compute usage might affect the rest of your cluster.
+Resource usage is more pertinent if you run resource-lite instances, as the UF would take up a more significant percentage of CPU and memory.
+Note that you can always increase CPU and memory levels for your cluster.
+
+### Splunk HTTP Event Collector
+
+The Splunk HTTP Event Collector (HEC) allows us to submit metrics via HTTP.
+We maintain a library [TelemetryMetricsSplunk](https://github.com/anthonyshull/telemetry_metrics_splunk) that forwards telemetry metrics to Splunk via the HEC.
+
+To use it, create a metrics index in Splunk and add its HEC token to the library configuration.
+
+```elixir
+config :app, TelemetryMetricsSplunk,
+  index: "your-index"
+  token: System.get_env("SPLUNK_TOKEN")
+```
+
+#### When you should choose this method
+
+- You only need metrics from Elixir applications.
+
+#### When you shouldn't choose this method
+
+- You maintain non-Elixir applications and do not want to write a method of getting metrics into Splunk.
+
+There aren't any libraries in other languages that transmit metrics to Splunk via the HEC.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+You can argue that we do not need metrics and that adding a means to collect them increases the complexity of our technology stack.
+However, we hope the limitations of a logs-only approach to observability have been made apparent by this RFC.
+Furthermore, we hope that the promised benefits of metrics cause excitement.
```

Comment:
suggestion: some other drawbacks offhand:
- additional resource usage for the Universal Forwarder
- need to keep the UF up-to-date
- possibly trickier to manage the connection to Splunk (vs. relying on ECS)

some questions:
- how do we monitor the UF?
- how does the UF handle being spun up / spun down elastically?

Reply:
I have addressed the resource usage above. It's a small footprint. Again, this is how Datadog et al. work. The UF has been around forever and is stable. Keeping it up-to-date would just be releasing a new image, which almost always only entails updating the version in the Dockerfile.

The UF isn't different from any other metrics agent; if you're worried that you aren't getting metrics, you write an alert that expects metrics to be sent. You don't need to scale the UF. You only need one.

Reply:
I thought the UF was implemented as a sidecar to Dotcom? So it's being spun up / spun down as Dotcom is.